security/rbac/engine/engine.go

```diff
@@ -0,0 +1,153 @@
+
+// Decision is the enum type that represents different authorization
+// decisions a CEL engine can return.
+type Decision int32
+
+const (
+	// DecisionAllow indicates allowing the RPC to go through.
+	DecisionAllow Decision = iota
+	// DecisionDeny indicates denying the RPC from going through.
+	DecisionDeny
+	// DecisionUnknown indicates that there is insufficient information to
+	// determine whether or not an RPC call is authorized.
+	DecisionUnknown
+)
+
+// String returns the string representation of a Decision object
+func (d Decision) String() string {
+	return [...]string{"DecisionAllow", "DecisionDeny", "DecisionUnknown"}[d]
+}
+
+// AuthorizationDecision is the output of CEL engine.
+type AuthorizationDecision struct {
+	decision   Decision
+	policyName string
+}
+
+// Converts an expression to a parsed expression, with SourceInfo nil
+func exprToParsedExpr(condition *expr.Expr) *expr.ParsedExpr {
+	return &expr.ParsedExpr{Expr: condition}
+}
+
+// Converts an expression to a CEL program
+func exprToProgram(condition *expr.Expr) *cel.Program {
+	return nil // TODO
+}
+
+// Returns whether or not a policy is matched.
+// If args is empty, the match always fails.
+func matches(program *cel.Program, args AuthorizationArgs) bool {
+	return false // TODO
+}
+
+// rbacEngine is the struct for an engine created from one RBAC proto.
+type rbacEngine struct {
+	action     pb.RBAC_Action
+	conditions map[string]*cel.Program
+}
+
+// Creates a new rbacEngine from an RBAC.
+func newRbacEngine(rbac pb.RBAC) rbacEngine {
+	action := rbac.Action
+	conditions := make(map[string]*cel.Program)
+	for policyName, policy := range rbac.Policies {
+		conditions[policyName] = exprToProgram(policy.Condition)
+	}
+	return rbacEngine{action, conditions}
+}
+
+// Returns whether a set of authorization arguments match an rbacEngine's conditions.
+func (engine rbacEngine) matches(args AuthorizationArgs) (bool, string) {
+	for policyName, condition := range engine.conditions {
+		if matches(condition, args) {
+			return true, policyName
+		}
+	}
+	return false, ""
+}
+
+// CelEvaluationEngine is the struct for CEL engine.
+type CelEvaluationEngine struct {
+	engines []rbacEngine
+}
+
+// NewCelEvaluationEngine builds a cel evaluation engine from a list of Envoy RBACs.
+func NewCelEvaluationEngine(rbacs []pb.RBAC) (CelEvaluationEngine, error) {
+	if len(rbacs) < 1 || len(rbacs) > 2 {
+		return CelEvaluationEngine{}, errWrongNumberRbacs
+	}
+	if len(rbacs) == 2 && (rbacs[0].Action != pb.RBAC_DENY || rbacs[1].Action != pb.RBAC_ALLOW) {
+		return CelEvaluationEngine{}, errWrongRbacActions
+	}
+	var engines []rbacEngine
+	for _, rbac := range rbacs {
+		engines = append(engines, newRbacEngine(rbac))
+	}
+	return CelEvaluationEngine{engines}, nil
+}
+
+// Evaluate is the core function that evaluates whether an RPC is authorized.
+func (celEngine CelEvaluationEngine) Evaluate(args AuthorizationArgs) AuthorizationDecision {
+	for _, engine := range celEngine.engines {
```

Comment:
The new authz logic is, 
1. we first check if there are policy matching. If there are, we just use whatever it is, since it is explicitly stated in the expression. For instance, if it is an ALLOW engine and we find a match, then we should allow it. 
2. If no match, we then try to find if there are any policies we couldn't check, which is the undecided case. For now let's just assume the cel evaluation will fail in those cases. If any of them is failing, we should return undecided, because we may consult other third party service for further prove. 
3. If no evaluation error nor matching(which simply means no matching at all), we will use default decision(so default for an ALLOW engine is DENY, default for DENY + ALLOW engines is DENY, etc). Let me know if you have any questions on this part and I am happy to explain it!
@cindyxue FYI. We might also want to have this in Java engine.
@jiangtaoli2016 Please correct me if I am wrong. Thanks!

Reply:
Let me clarify a little bit further. Engine will return UNDECIDED only if there are attributes missing such that CEL cannot evaluate, e.g., if CEL expression needs an attribute `connect.uri_san_peer_certificate`, gRPC stack does not provide such value (peer does not have a valid SPIFFE ID), thus it will be undecided. 

E.g., if CEL expression is `connect.uri_san_peer_certificate == 'foo' || source.address == '1.2.3.4'`. If peer has not have a valid uri_san_peer_certificate, but address is '1.2.3.4', then CEL will evaluate as `true`. If peer has not uri_san_peer_certificate and address is '1.1.1.1', it will be undecided for this particular CEL evaluation, because we do not have enough information to evaluate a conclusive decision.

Now, back to no matching rule
- If RBAC engine only has one ALLOW policy, then no matching means DENY
- If RBAC engine only has one DENY policy, then no matching means ALLOW
- If RBAC engine has DENY + ALLOW, then no matching means DENY

I will need to update design next week to clarify these.